src/lua/datetime.lua

```diff
@@ -56,6 +56,16 @@ int     tnt_dt_month        (dt_t dt);
 int     tnt_dt_doy          (dt_t dt);
 int     tnt_dt_dom          (dt_t dt);
 
+/* dt_arithmetic.h definitions */
+typedef enum {
+    DT_EXCESS,
+    DT_LIMIT,
+    DT_SNAP
+} dt_adjust_t;
```

Comment:
@Gerold103 said:
> Oh, sorry, one another thing - the RFC was updated. So the last month day snap is now not the default behaviour. 
> We must drop it for now. @Mons promised to come with more details about the option.

We now export these 3 options (`DT_EXCESS`, `DT_LIMIT`, `DT_SNAP`) as separate options available as last argument in `date:add()` or `date:sub()`. Please see spec https://hackmd.io/@Mons/S1Vfc_axK#%D0%9F%D0%B0%D1%80%D0%B0%D0%BC%D0%B5%D1%82%D1%80-adjust

```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -663,8 +663,14 @@ local function interval_tostring(o)
     return s
 end
 
+local adjust_xlat = {
+    none = builtin.DT_LIMIT,
+    last = builtin.DT_SNAP,
+    excess = builtin.DT_EXCESS,
+}
+
 local function datetime_increment_by(self, direction, years, months,
-                                     seconds, nanoseconds)
+                                     seconds, nanoseconds, adjust)
     -- operations with intervals should be done using local human dates
     -- not UTC dates, thus normalize to local timezone before operations.
     local dt = local_dt(self)
@@ -672,18 +678,25 @@ local function datetime_increment_by(self, direction, years, months,
     local offset = self.tzoffset
 
     local is_ym_updated = false
+    if adjust ~= nil then
+        local operation = direction >= 0 and 'addition' or 'subtraction'
+        check_table(adjust, operation)
+        adjust = adjust_xlat[adjust.adjust] or builtin.DT_SNAP
+    else
+        adjust = builtin.DT_SNAP
+    end
 
     if years ~= 0 then
         check_range(years, MIN_DATE_YEAR, MAX_DATE_YEAR, 'years')
         local approx_dt = dt + direction * years * AVERAGE_DAYS_YEAR
         check_dt(approx_dt, direction, years, 'years')
-        dt = builtin.tnt_dt_add_years(dt, direction * years, builtin.DT_SNAP)
+        dt = builtin.tnt_dt_add_years(dt, direction * years, adjust)
         is_ym_updated = true
     end
     if months ~= 0 then
         local new_dt = dt + direction * months * AVERAGE_DAYS_MONTH
         check_dt(new_dt, direction, months, 'months')
-        dt = builtin.tnt_dt_add_months(dt, direction * months, builtin.DT_SNAP)
+        dt = builtin.tnt_dt_add_months(dt, direction * months, adjust)
         is_ym_updated = true
     end
     if is_ym_updated then
@@ -833,13 +846,13 @@ end
 
 -- addition or subtraction from date/time of a given interval
 -- described via table direction should be +1 or -1
-local function datetime_shift(self, o, direction)
+local function datetime_shift(self, o, direction, adjust)
     assert(direction == -1 or direction == 1)
     local title = direction > 0 and "datetime.add" or "datetime.sub"
     check_interval_table(o, title)
 
     local year, month, secs, nsec = interval_decouple_args(o)
-    return datetime_increment_by(self, direction, year, month, secs, nsec)
+    return datetime_increment_by(self, direction, year, month, secs, nsec, adjust)
 end
 
 --[[
@@ -1077,8 +1090,12 @@ local datetime_index_functions = {
     format = datetime_format,
     totable = datetime_totable,
     set = datetime_set,
-    add = function(self, obj) return datetime_shift(self, obj, 1) end,
-    sub = function(self, obj) return datetime_shift(self, obj, -1) end,
+    add = function(self, obj, adjust)
+            return datetime_shift(self, obj, 1, adjust)
+          end,
+    sub = function(self, obj, adjust)
+            return datetime_shift(self, obj, -1, adjust)
+          end,
 }
 
 local function datetime_index(self, key)
```

This change required significant update to test scripts:

<details>
  <summary>Click to expand...</summary>

```diff
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -5,7 +5,7 @@ local test = tap.test('errno')
 local date = require('datetime')
 local ffi = require('ffi')
 
-test:plan(22)
+test:plan(23)
 
 -- minimum supported date - -5879610-06-22
 local MIN_DATE_YEAR = -5879610
@@ -653,57 +653,68 @@ test:test("Time interval __index fields", function(test)
 end)
 
 test:test("Time interval operations", function(test)
-    test:plan(114)
+    test:plan(162)
 
     local ival_new = date.interval.new
 
     -- check arithmetic with leap dates
     local base_leap_date = { year = 1972, month = 2, day = 29}
-    local tadd = date.new(base_leap_date)
-    local tsub = date.new(base_leap_date)
+
     local leap_ivals_check = {
-        {{year = 1, month = 2}, '1973-05-01T00:00:00Z', '1971-01-01T00:00:00Z'},
-        {{year = 2, month = 3}, '1975-08-01T00:00:00Z', '1968-10-01T00:00:00Z'},
-        {{year = -1}, '1974-08-01T00:00:00Z', '1969-10-02T00:00:00Z'},
-        {{year = -1}, '1973-08-01T00:00:00Z', '1970-10-02T00:00:00Z'},
+        {{month = 1}, '1972-03-31T00:00:00Z', '1972-01-31T00:00:00Z'},
+        {{month = 1}, '1972-04-30T00:00:00Z', '1971-12-31T00:00:00Z'},
+        {{month = 5}, '1972-09-30T00:00:00Z', '1971-07-31T00:00:00Z'},
+        {{year = 1, month = 2}, '1973-12-01T00:00:00Z', '1970-05-31T00:00:00Z'},
+        {{year = 2, month = 3}, '1976-03-01T00:00:00Z', '1968-02-29T00:00:00Z'},
+        {{month = 1}, '1976-04-01T00:00:00Z', '1968-01-31T00:00:00Z'},
+        {{month = 1}, '1976-05-01T00:00:00Z', '1967-12-31T00:00:00Z'},
+        {{month = 5}, '1976-10-01T00:00:00Z', '1967-07-31T00:00:00Z'},
+        {{year = -1}, '1975-10-02T00:00:00Z', '1968-07-30T00:00:00Z'},
+        {{year = -1}, '1974-10-02T00:00:00Z', '1969-07-31T00:00:00Z'},
     }
-    -- check :add{table}
+
+    local tadd_t = date.new(base_leap_date) -- :add{table}
+    local tadd_i = date.new(base_leap_date) -- :add{interval}
+    local tplus_i = date.new(base_leap_date) -- t + interval
+    local tplus_t = date.new(base_leap_date) -- t + {table}
+
+    -- check additions
     for _, row in pairs(leap_ivals_check) do
-        local delta, sadd, ssub = unpack(row)
-        test:is(tostring(tadd:add(delta)), sadd,
+        local delta, sadd, _ = unpack(row)
+        -- t:add{table}
+        test:is(tostring(tadd_t:add(delta)), sadd,
                 ('add delta to %s'):format(sadd))
-        test:is(tostring(tsub:sub(delta)), ssub,
-                ('sub delta to %s'):format(ssub))
-    end
-    -- check :add{interval}
-    tadd = date.new(base_leap_date)
-    tsub = date.new(base_leap_date)
-    for _, row in pairs(leap_ivals_check) do
-        local delta, sadd, ssub = unpack(row)
-        test:is(tostring(tadd:add(ival_new(delta))), sadd,
+        -- t:add(interval)
+        test:is(tostring(tadd_i:add(ival_new(delta))), sadd,
                 ('add ival delta to %s'):format(sadd))
-        test:is(tostring(tsub:sub(ival_new(delta))), ssub,
-                ('sub ival delta to %s'):format(ssub))
-    end
-    -- check t + interval
-    tadd = date.new(base_leap_date)
-    tsub = date.new(base_leap_date)
-    for _, row in pairs(leap_ivals_check) do
-        local delta, sadd, ssub = unpack(row)
-        tadd = tadd + ival_new(delta)
-        test:is(tostring(tadd), sadd, ('%s: t + ival delta'):format(sadd))
-        tsub = tsub - ival_new(delta)
-        test:is(tostring(tsub), ssub, ('%s: t - ival delta'):format(ssub))
+        -- t + interval
+        tplus_i = tplus_i + ival_new(delta)
+        test:is(tostring(tplus_i), sadd, ('%s: t + ival delta'):format(sadd))
+        -- t - interval
+        tplus_t = tplus_t + delta
+        test:is(tostring(tplus_t), sadd, ('%s: t + table delta'):format(sadd))
     end
-    -- check t + {table}
-    tadd = date.new(base_leap_date)
-    tsub = date.new(base_leap_date)
+
+    local tsub_t = date.new(base_leap_date) -- :sub{table}
+    local tsub_i = date.new(base_leap_date) -- :sub{interval}
+    local tminus_i = date.new(base_leap_date) -- t - interval
+    local tminus_t = date.new(base_leap_date) -- t - {table}
+
+    -- check subtractions
     for _, row in pairs(leap_ivals_check) do
-        local delta, sadd, ssub = unpack(row)
-        tadd = tadd + delta
-        test:is(tostring(tadd), sadd, ('%s: t + table delta'):format(sadd))
-        tsub = tsub - delta
-        test:is(tostring(tsub), ssub, ('%s: t - table delta'):format(ssub))
+        local delta, _, ssub = unpack(row)
+        -- t:add{table}
+        test:is(tostring(tsub_t:sub(delta)), ssub,
+                ('sub delta to %s'):format(ssub))
+        -- t:add(interval)
+        test:is(tostring(tsub_i:sub(ival_new(delta))), ssub,
+                ('sub ival delta to %s'):format(ssub))
+        -- t + interval
+        tminus_i = tminus_i - ival_new(delta)
+        test:is(tostring(tminus_i), ssub, ('%s: t - ival delta'):format(ssub))
+        -- t - interval
+        tminus_t = tminus_t - delta
+        test:is(tostring(tminus_t), ssub, ('%s: t - table delta'):format(ssub))
     end
 
     -- check average, not leap dates
@@ -722,48 +733,48 @@ test:test("Time interval operations", function(test)
          '1968-08-13T23:27:05.876544Z'},
     }
 
-    -- check :add{table}
-    tadd = date.new(base_non_leap_date)
-    tsub = date.new(base_non_leap_date)
-    for _, row in pairs(non_leap_ivals_check) do
-        local delta, sadd, ssub = unpack(row)
-        test:is(tostring(tadd:add(delta)), sadd,
-                ('add delta to %s'):format(sadd))
-        test:is(tostring(tsub:sub(delta)), ssub,
-                ('sub delta to %s'):format(ssub))
-    end
+    tadd_t = date.new(base_non_leap_date) -- :add{table}
+    tadd_i = date.new(base_non_leap_date) -- :add{interval}
+    tplus_i = date.new(base_non_leap_date) -- t + interval
+    tplus_t = date.new(base_non_leap_date) -- t + {table}
 
-    -- check :add{interval}
-    tadd = date.new(base_non_leap_date)
-    tsub = date.new(base_non_leap_date)
+    -- check additions
     for _, row in pairs(non_leap_ivals_check) do
-        local delta, sadd, ssub = unpack(row)
-        test:is(tostring(tadd:add(ival_new(delta))), sadd,
+        local delta, sadd, _ = unpack(row)
+        -- t:add{table}
+        test:is(tostring(tadd_t:add(delta)), sadd,
                 ('add delta to %s'):format(sadd))
-        test:is(tostring(tsub:sub(ival_new(delta))), ssub,
-                ('sub delta to %s'):format(ssub))
+        -- t:add(interval)
+        test:is(tostring(tadd_i:add(ival_new(delta))), sadd,
+                ('add ival delta to %s'):format(sadd))
+        -- t + interval
+        tplus_i = tplus_i + ival_new(delta)
+        test:is(tostring(tplus_i), sadd, ('%s: t + ival delta'):format(sadd))
+        -- t - interval
+        tplus_t = tplus_t + delta
+        test:is(tostring(tplus_t), sadd, ('%s: t + table delta'):format(sadd))
     end
 
-    -- check t + interval
-    tadd = date.new(base_non_leap_date)
-    tsub = date.new(base_non_leap_date)
-    for _, row in pairs(non_leap_ivals_check) do
-        local delta, sadd, ssub = unpack(row)
-        tadd = tadd + ival_new(delta)
-        test:is(tostring(tadd), sadd, ('%s: t + delta'):format(sadd))
-        tsub = tsub - ival_new(delta)
-        test:is(tostring(tsub), ssub, ('%s: t - delta'):format(ssub))
-    end
+    tsub_t = date.new(base_non_leap_date) -- :sub{table}
+    tsub_i = date.new(base_non_leap_date) -- :sub{interval}
+    tminus_i = date.new(base_non_leap_date) -- t - interval
+    tminus_t = date.new(base_non_leap_date) -- t - {table}
 
-    -- check t + {table}
-    tadd = date.new(base_non_leap_date)
-    tsub = date.new(base_non_leap_date)
+    -- check subtractions
     for _, row in pairs(non_leap_ivals_check) do
-        local delta, sadd, ssub = unpack(row)
-        tadd = tadd + delta
-        test:is(tostring(tadd), sadd, ('%s: t + delta'):format(sadd))
-        tsub = tsub - delta
-        test:is(tostring(tsub), ssub, ('%s: t - delta'):format(ssub))
+        local delta, _, ssub = unpack(row)
+        -- t:add{table}
+        test:is(tostring(tsub_t:sub(delta)), ssub,
+                ('sub delta to %s'):format(ssub))
+        -- t:add(interval)
+        test:is(tostring(tsub_i:sub(ival_new(delta))), ssub,
+                ('sub ival delta to %s'):format(ssub))
+        -- t + interval
+        tminus_i = tminus_i - ival_new(delta)
+        test:is(tostring(tminus_i), ssub, ('%s: t - ival delta'):format(ssub))
+        -- t - interval
+        tminus_t = tminus_t - delta
+        test:is(tostring(tminus_t), ssub, ('%s: t - table delta'):format(ssub))
     end
 
     local specific_errors = {
@@ -777,13 +788,104 @@ test:test("Time interval operations", function(test)
     }
     for _, row in pairs(specific_errors) do
         local err_msg, obj = unpack(row)
-        assert_raises(test, err_msg, function() return tadd:add(obj) end)
-        assert_raises(test, err_msg, function() return tsub:sub(obj) end)
+        assert_raises(test, err_msg, function() return tadd_t:add(obj) end)
+        assert_raises(test, err_msg, function() return tsub_t:sub(obj) end)
+    end
+    assert_raises_like(test, expected_interval_but, function() tadd_t:add('bogus') end)
+    assert_raises_like(test, expected_interval_but, function() tadd_t:add(123) end)
+    assert_raises_like(test, expected_interval_but, function() tsub_t:sub('bogus') end)
+    assert_raises_like(test, expected_interval_but, function() tsub_t:sub(123) end)
+end)
+
+test:test("Time interval operations - different adjustments", function(test)
+    test:plan(60)
+
+    -- check arithmetic with leap dates
+    local base_leap_date = { year = 1972, month = 2, day = 29}
+    local adj_modes = {
+        [1] = "last", -- default if omitted
+        [2] = "none",
+        [3] = "excess",
+    }
+
+    local leap_ivals_check_add = {
+        {{month = 1}, {'1972-03-31T00:00:00Z', '1972-03-29T00:00:00Z',
+                       '1972-03-29T00:00:00Z'}},
+        {{month = 1}, {'1972-04-30T00:00:00Z', '1972-04-29T00:00:00Z',
+                       '1972-04-29T00:00:00Z'}},
+        {{month = 5}, {'1972-09-30T00:00:00Z', '1972-09-29T00:00:00Z',
+                       '1972-09-29T00:00:00Z'}},
+        {{year = 1, month = 2}, {'1973-12-01T00:00:00Z', '1973-11-30T00:00:00Z',
+                                 '1973-11-30T00:00:00Z'}},
+        {{year = 2, month = 3}, {'1976-03-01T00:00:00Z', '1976-02-29T00:00:00Z',
+                                 '1976-03-01T00:00:00Z'}},
+        {{month = 1}, {'1976-04-01T00:00:00Z', '1976-03-29T00:00:00Z',
+                       '1976-04-01T00:00:00Z'}},
+        {{month = 1}, {'1976-05-01T00:00:00Z', '1976-04-29T00:00:00Z',
+                       '1976-05-01T00:00:00Z'}},
+        {{month = 5}, {'1976-10-01T00:00:00Z', '1976-09-29T00:00:00Z',
+                       '1976-10-01T00:00:00Z'}},
+        {{year = -1}, {'1975-10-02T00:00:00Z', '1975-09-30T00:00:00Z',
+                       '1975-10-02T00:00:00Z'}},
+        {{year = -1}, {'1974-10-02T00:00:00Z', '1974-09-30T00:00:00Z',
+                       '1974-10-02T00:00:00Z'}},
+    }
+
+    local tadd = {}
+    for i=1,3 do
+        tadd[i] = date.new(base_leap_date)
+    end
+
+    -- check additions with adjustments
+    for _, row in pairs(leap_ivals_check_add) do
+        local delta, sadds = unpack(row)
+        for i=1,3 do
+            local adjust = adj_modes[i]
+            local sadd = sadds[i]
+            test:is(tostring(tadd[i]:add(delta, {adjust=adjust})), sadd,
+                    ('add delta to %s with adjust %s'):format(sadd, adjust))
+
+        end
+    end
+
+    local leap_ivals_check_sub = {
+        {{month = 1}, {'1972-01-31T00:00:00Z', '1972-01-29T00:00:00Z',
+                       '1972-01-29T00:00:00Z'}},
+        {{month = 1}, {'1971-12-31T00:00:00Z', '1971-12-29T00:00:00Z',
+                       '1971-12-29T00:00:00Z'}},
+        {{month = 5}, {'1971-07-31T00:00:00Z', '1971-07-29T00:00:00Z',
+                       '1971-07-29T00:00:00Z'}},
+        {{year = 1, month = 2}, {'1970-05-31T00:00:00Z', '1970-05-29T00:00:00Z',
+                       '1970-05-29T00:00:00Z'}},
+        {{year = 2, month = 3}, {'1968-02-29T00:00:00Z', '1968-02-28T00:00:00Z',
+                       '1968-02-28T00:00:00Z'}},
+        {{month = 1}, {'1968-01-31T00:00:00Z', '1968-01-28T00:00:00Z',
+                       '1968-01-28T00:00:00Z'}},
+        {{month = 1}, {'1967-12-31T00:00:00Z', '1967-12-28T00:00:00Z',
+                       '1967-12-28T00:00:00Z'}},
+        {{month = 5}, {'1967-07-31T00:00:00Z', '1967-07-28T00:00:00Z',
+                       '1967-07-28T00:00:00Z'}},
+        {{year = -1}, {'1968-07-30T00:00:00Z', '1968-07-27T00:00:00Z',
+                       '1968-07-27T00:00:00Z'}},
+        {{year = -1}, {'1969-07-31T00:00:00Z', '1969-07-28T00:00:00Z',
+                       '1969-07-28T00:00:00Z'}},
+    }
+
+    local tsub = {}
+    for i=1,3 do
+        tsub[i] = date.new(base_leap_date)
+    end
+
+    -- check subtractions with adjustments
+    for _, row in pairs(leap_ivals_check_sub) do
+        local delta, ssubs = unpack(row)
+        for i=1,3 do
+            local adjust = adj_modes[i]
+            local ssub = ssubs[i]
+            test:is(tostring(tsub[i]:sub(delta, {adjust=adjust})), ssub,
+                    ('sub delta to %s with adjust %s'):format(ssub, adjust))
+        end
     end
-    assert_raises_like(test, expected_interval_but, function() tadd:add('bogus') end)
-    assert_raises_like(test, expected_interval_but, function() tadd:add(123) end)
-    assert_raises_like(test, expected_interval_but, function() tsub:sub('bogus') end)
-    assert_raises_like(test, expected_interval_but, function() tsub:sub(123) end)
 end)
 
 test:test("Time intervals creation - range checks", function(test)
```
</details>

Reply:
Good point, indeed. `DT_SNAP` was default before we introduced `adjust` parameter. Will switch to `none` (`DT_LIMIT`) as default if there is no adjust parameter passed.

Reply:
Please see (slightly incorrect patch) which fixed defautlt.

```diff
commit a93447f999a7757d5478e3fb5ccec02a8c359094
Author: Timur Safin <tsafin@tarantool.org>
Date:   Mon Nov 29 18:09:36 2021 +0300

    temp - serge petrenko comment - default is 'none' not 'last'

diff --git a/src/lua/datetime.lua b/src/lua/datetime.lua
index cec27aa16..fbd250a4b 100644
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -681,9 +681,9 @@ local function datetime_increment_by(self, direction, years, months,
     if adjust ~= nil then
         local operation = direction >= 0 and 'addition' or 'subtraction'
         check_table(adjust, operation)
-        adjust = adjust_xlat[adjust.adjust] or builtin.DT_SNAP
+        adjust = adjust_xlat[adjust.adjust] or builtin.DT_LIMIT
     else
-        adjust = builtin.DT_SNAP
+        adjust = builtin.DT_LIMIT
     end
 
     if years ~= 0 then
diff --git a/test/app-tap/datetime.test.lua b/test/app-tap/datetime.test.lua
index e6377bd6c..826a02595 100755
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -661,16 +661,16 @@ test:test("Time interval operations", function(test)
     local base_leap_date = { year = 1972, month = 2, day = 29}
 
     local leap_ivals_check = {
-        {{month = 1}, '1972-03-31T00:00:00Z', '1972-01-31T00:00:00Z'},
-        {{month = 1}, '1972-04-30T00:00:00Z', '1971-12-31T00:00:00Z'},
-        {{month = 5}, '1972-09-30T00:00:00Z', '1971-07-31T00:00:00Z'},
-        {{year = 1, month = 2}, '1973-12-01T00:00:00Z', '1970-05-31T00:00:00Z'},
-        {{year = 2, month = 3}, '1976-03-01T00:00:00Z', '1968-02-29T00:00:00Z'},
-        {{month = 1}, '1976-04-01T00:00:00Z', '1968-01-31T00:00:00Z'},
-        {{month = 1}, '1976-05-01T00:00:00Z', '1967-12-31T00:00:00Z'},
-        {{month = 5}, '1976-10-01T00:00:00Z', '1967-07-31T00:00:00Z'},
-        {{year = -1}, '1975-10-02T00:00:00Z', '1968-07-30T00:00:00Z'},
-        {{year = -1}, '1974-10-02T00:00:00Z', '1969-07-31T00:00:00Z'},
+        {{month = 1}, '1972-03-29T00:00:00Z', '1972-01-29T00:00:00Z'},
+        {{month = 1}, '1972-04-29T00:00:00Z', '1971-12-29T00:00:00Z'},
+        {{month = 5}, '1972-09-29T00:00:00Z', '1971-07-29T00:00:00Z'},
+        {{year = 1, month = 2}, '1973-11-30T00:00:00Z', '1970-05-29T00:00:00Z'},
+        {{year = 2, month = 3}, '1976-02-29T00:00:00Z', '1968-02-28T00:00:00Z'},
+        {{month = 1}, '1976-03-29T00:00:00Z', '1968-01-28T00:00:00Z'},
+        {{month = 1}, '1976-04-29T00:00:00Z', '1967-12-28T00:00:00Z'},
+        {{month = 5}, '1976-09-29T00:00:00Z', '1967-07-28T00:00:00Z'},
+        {{year = -1}, '1975-09-30T00:00:00Z', '1968-07-27T00:00:00Z'},
+        {{year = -1}, '1974-09-30T00:00:00Z', '1969-07-28T00:00:00Z'},
     }
 
     local tadd_t = date.new(base_leap_date) -- :add{table}
```

> If we carefully review series of dates we step over after additions/subtractions we could see some wrong dates visited. That's exactly the problem you say below, please see final fix below.

Reply:
Yeah, I've edited commit message and now it's saying:

    If one needs a way of arithmetics proceeded for day of months
    then there is extra optional argument in `:add()` or `:sub()`
    methods, which allow control round of day in month after
    arithmetic operation:
    
    - 'last' - mode when day snaps to the end of month, if happens
    
      -- 29.02.2004 -> 31.03.2001
      dt:add( {month = 1, adjust = "last" } )
    
    - 'none' only tuncation proceeded (default mode)
    
      -- 29.02.* -> 29.03.*
      dt:add( {month = 1, adjust = "none" } )
    
    - 'excess' - overflow mode, without any snap or truncation to
      the end of month

Reply:
Typo in the commit message:
>    - 'none' only tuncation ito the end of month proceeded
      (**default mode**):
    
ito -> to